crates/wasi/src/preview2/pipe.rs

```diff
@@ -74,42 +78,55 @@ impl MemoryOutputPipe {
 
 #[async_trait::async_trait]
 impl HostOutputStream for MemoryOutputPipe {
-    fn write(&mut self, bytes: Bytes) -> Result<(usize, StreamState), anyhow::Error> {
+    fn write(&mut self, bytes: Bytes) -> Result<(), OutputStreamError> {
         let mut buf = self.buffer.lock().unwrap();
+        if bytes.len() > self.capacity - buf.len() {
+            return Err(OutputStreamError::Trap(anyhow!(
+                "write beyond capacity of MemoryOutputPipe"
+            )));
+        }
         buf.extend_from_slice(bytes.as_ref());
-        Ok((bytes.len(), StreamState::Open))
+        // Always ready for writing
+        Ok(())
     }
-
-    async fn ready(&mut self) -> Result<(), Error> {
-        // This stream is always ready for writing.
+    fn flush(&mut self) -> Result<(), OutputStreamError> {
+        // This stream is always flushed
         Ok(())
     }
+    async fn write_ready(&mut self) -> Result<usize, OutputStreamError> {
+        let consumed = self.buffer.lock().unwrap().len();
+        if consumed < self.capacity {
+            Ok(self.capacity - consumed)
+        } else {
+            Ok(0)
+        }
+    }
```

Comment:
Question on this, is the intention that the host will consume output from a `MemoryOutputBuffer` incrementally, freeing up space in the pipe? Or is the intention that everything is collected and then the pipe is completely done and no more data can ever be put in?

Reply:
The intention is that writes get collected up to the stated capacity, then no more are accepted.

We could return Err(OutputStreamError::Closed) once capacity is reached, but I wasn't really sure.

We basically only use this type to collect output in tests.